processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import io.druid.java.util.common.IAE;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.segment.data.IndexedInts;
+import io.druid.segment.filter.BooleanValueMatcher;
+
+import java.util.BitSet;
+import java.util.Objects;
+
+public final class DimensionSelectorUtils
+{
+
+  private DimensionSelectorUtils()
+  {
+  }
+
+  public static ValueMatcher makeRowBasedValueMatcher(
+      DimensionSelector selector,
+      String value,
+      boolean matchNull
+  )
+  {
+    IdLookup idLookup = selector.idLookup();
```

Comment:
IMO it doens't make sense because the "entry point" to `DimensionSelector.makeValueMatcher()` is `StringValueMatcherColumnSelectorStrategy`. This optimization is applied here. And `DimensionSelectorUtils`'s methods are called _from_ `DimensionSelector` implementations.

Also I removed `(cardinality == 1 && selector.lookupName(0) == null)` part in `StringValueMatcherColumnSelectorStrategy`, because it really means "this DimensionSelector is `NullDimensionSelector`", so delegate to `NullDimensionSelector` which also returns only `BooleanValueMatcher`.